snooty/rstparser.py

```diff
@@ -581,18 +643,79 @@ def add_diagnostics(self, diagnostics: Iterable[Diagnostic]) -> None:
 _V = TypeVar("_V", bound=Visitor)
 
 
+@dataclass
+class Domain:
+    directives: Dict[str, Type[Any]] = field(default_factory=dict)
+    roles: Dict[str, RoleHandler] = field(default_factory=dict)
+
+
+@dataclass
+class Registry:
+    domains: DefaultDict[str, Domain] = field(
+        default_factory=lambda: defaultdict(Domain)
+    )
+
+    def add_directive(self, name: str, directive: Type[Any]) -> None:
+        domain, name = util.split_domain(name)
+        self.domains[domain].directives[name] = directive
+
+    def add_role(self, name: str, role: RoleHandler) -> None:
+        domain, name = util.split_domain(name)
+        self.domains[domain].roles[name] = role
+
+    def lookup_directive(
+        self,
+        directive_name: str,
+        language_module: object,
+        document: docutils.nodes.document,
+    ) -> Tuple[Optional[Type[Any]], List[object]]:
+        # Remove the built-in directives we don't want
+        domain_name, directive_name = util.split_domain(directive_name)
+        if domain_name:
+            return self.domains[domain_name].directives.get(directive_name, None), []
+
+        for domain_name in ["mongodb", "std", ""]:
```

Comment:
I know you mentioned this in the PR description, but what exactly is going on here?

Reply:
Moved this list into a constant called `DOMAIN_RESOLUTION_SEQUENCE` with an explanatory comment